File: examples/nodejs/_3_Object/publisher.js
const SharedPubSub = require('SharedPubSub');
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const publisher = new SharedPubSub.Publisher_ExampleClass("Object");

async function loop() {
    while (true) {
        /*--------- Example 1 : Publish  ---------*/
        let exampleObject = publisher.readValue();
        exampleObject.value1+=1;
        exampleObject.value2+=0.1;
        exampleObject.value3+=10;
        publisher.publish(exampleObject);
        exampleObject.printValues();
        await sleep(1000)
        /*--------- Example 2 : Get raw pointer, change value and notify  ---------*/
        // Although it's possible, we will not do raw pointer examples
    }
}

loop().catch(console.error);

File: examples/cpp/_4_ObjectAtomic/publisher.cpp
#include <iostream>
#include <thread>
#include "SharedPubSub.hpp"
#include "Examples.h"
using namespace std;

int main(){
    
    shps::Publisher<ExampleClassAtomic> publisher("ObjectAtomic");
    ExampleClassAtomic exampleObjectAtomic;
    while(1){
        /*--------- Example 1 : Publish  ---------*/
        exampleObjectAtomic = publisher.readValue();
        exampleObjectAtomic.value1+=1;
        exampleObjectAtomic.value2+=0.1;
        exampleObjectAtomic.value3+=10;
        publisher.publish(exampleObjectAtomic);
        exampleObjectAtomic.printValues();

        this_thread::sleep_for(1s);

        /*--------- Example 2 : Get raw pointer, change value and notify  ---------*/
        ExampleClassAtomic* rawExampleObjectAtomic = publisher.rawValue();
        rawExampleObjectAtomic->value1+=1;
        rawExampleObjectAtomic->value2+=0.1;
        rawExampleObjectAtomic->value3+=10;
        publisher.notifyAll();
        rawExampleObjectAtomic->printValues();
        this_thread::sleep_for(1s);
    }
    return 0;
}